### PR TITLE
[ALLUXIO-3362] Move inode locking out of inode tree

### PR DESCRIPTION
**Pull request by aaudiber**
https://alluxio.atlassian.net/browse/ALLUXIO-3362

This PR reduces the memory cost of inode locks. Instead of always storing a lock for every inode, we create inode locks only when needed, and GC them when they are no longer in use.

### CODE REVIEW

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeLocker.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+/**
+ * Class for managing inode locking.
+ */
+public class InodeLocker {
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeLocker.java]: NIT: can we name this `InodeLockManager`?

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeLockList.java [hunk 2]
```diff
@@ -162,9 +192,31 @@ public synchronized void lockWriteAndCheckParent(InodeView inode, InodeView pare
...
+    lockWriteAndCheckParent(inode, parent);
+    try {
+      checkName(inode, name);
+    } catch (Throwable t) {
+      unlockLast();
+      throw t;
+    }
+  }
+
+  private void checkExists(InodeView inode) throws InvalidPathException {
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeLockList.java]: why are these 3 checks not part of `Inode` (or `InodeView`)?

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java [hunk 3]
```diff
@@ -633,8 +643,7 @@ public CreatePathResult createPath(RpcContext rpcContext, LockedInodePath inodeP
             mDirectoryIdGenerator.getNewDirectoryId(rpcContext.getJournalContext()),
             currentInodeDirectory.getId(), pathComponents[k], missingDirOptions);
         // Lock the newly created inode before subsequent operations, and add it to the lock group.
-        extensibleInodePath.getLockList().lockWriteAndCheckNameAndParent(newDir,
-            currentInodeDirectory, pathComponents[k]);
+        extensibleInodePath.getLockList().lockWrite(newDir);
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java]: why no more checks?

**Comment by aaudiber** [comment]
@gpang @yuzhu FYI I'm working on a larger locking change now where we lock both nodes and edges. Edge locking simplifies a lot, especially when inodes could potentially be stored externally. Many of the changes here will still apply, but I suggest waiting until tomorrow for before diving into more detailed review.

**Comment by aaudiber** [comment]
re-opening in https://github.com/Alluxio/alluxio/pull/8199

**Pull request closed without merging**